src/kvstore/Part.cpp

```diff
@@ -5,34 +5,282 @@
  */
 
 #include "kvstore/Part.h"
+#include "wal/BufferFlusher.h"
+
+DEFINE_int32(cluster_id, 0, "A unique id for each cluster");
 
 namespace nebula {
 namespace kvstore {
 
-void SimplePart::asyncMultiPut(std::vector<KV> keyValues, KVCallback cb) {
-    CHECK_NOTNULL(engine_);
-    auto ret = engine_->multiPut(std::move(keyValues));
-    cb(ret, HostAddr(0, 0));
+namespace {
+
+static const char OP_PUT            = 0x1;
+static const char OP_MULTI_PUT      = 0x2;
+static const char OP_REMOVE         = 0x3;
+static const char OP_MULTI_REMOVE   = 0x4;
+static const char OP_REMOVE_PREFIX  = 0x5;
+static const char OP_REMOVE_RANGE   = 0x6;
+
+wal::BufferFlusher* getBufferFlusher() {
+    static wal::BufferFlusher flusher;
+    return &flusher;
+}
+
+}  // Anonymous namespace
+
+
+Part::Part(GraphSpaceID spaceId,
+           PartitionID partId,
+           HostAddr localAddr,
+           const std::string& walPath,
+           KVEngine* engine,
+           std::shared_ptr<folly::IOThreadPoolExecutor> ioPool,
+           std::shared_ptr<thread::GenericThreadPool> workers)
+        : RaftPart(FLAGS_cluster_id,
+                   spaceId,
+                   partId,
+                   localAddr,
+                   walPath,
+                   getBufferFlusher(),
+                   ioPool,
+                   workers)
+        , spaceId_(spaceId)
+        , partId_(partId)
+        , walPath_(walPath)
+        , engine_(engine) {
+}
+
+
+void Part::asyncPut(folly::StringPiece key, folly::StringPiece value, KVCallback cb) {
+    std::string log;
+    log.reserve(key.size() + value.size() + 9);
+    log.append(&OP_PUT, 1);
+    uint32_t len = static_cast<uint32_t>(key.size());
+    log.append(reinterpret_cast<char*>(&len), sizeof(len));
+    log.append(key.begin(), key.size());
+    len = static_cast<uint32_t>(value.size());
+    log.append(reinterpret_cast<char*>(&len), sizeof(len));
+    log.append(value.begin(), value.size());
+
+    appendAsync(FLAGS_cluster_id, std::move(log))
+        .then([callback = std::move(cb)] (AppendLogResult res) {
+            if (res == AppendLogResult::SUCCEEDED) {
```

Comment:
I'm kinda agree with @dutor . We need to clearly define what **duplicated code** means. I don't think a few lines of same code is duplicate. To me, **duplicated code** means a big chunk of code doing similar things

Previously we agreed to avoid using macro. I think "exiting a function from inside a macro" is especially bad, because a macro should not bring surprise/side effects into the code